// speechmatics/dist/batch/client.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var client_exports = {};
__export(client_exports, {
  BatchTranscription: () => BatchTranscription
});
module.exports = __toCommonJS(client_exports);
var import_connection = require("../config/connection");
var import_request = require("../utils/request");
var import_poll = __toESM(require("../utils/poll"));
class BatchTranscription {
  constructor(config) {
    this._apiKey = void 0;
    this.config = new import_connection.ConnectionConfigFull(config);
  }
  get apiKey() {
    return this._apiKey ?? (typeof this.config.apiKey === "string" ? this.config.apiKey : void 0);
  }
  async refreshApiKey() {
    if (typeof this.config.apiKey === "string")
      return this.config.apiKey;
    const newKey = await this.config.apiKey();
    this._apiKey = newKey;
    return newKey;
  }
  async refreshOnFail(doRequest) {
    try {
      return await doRequest(this.apiKey ?? await this.refreshApiKey());
    } catch (e) {
      console.info("Retrying due to error:", e);
      return await doRequest(await this.refreshApiKey());
    }
  }
  async get(endpoint, contentType, queryParams) {
    return await this.refreshOnFail(
      (key) => (0, import_request.request)(
        key,
        this.config.batchUrl,
        endpoint,
        "GET",
        null,
        { ...queryParams, [import_request.SM_APP_PARAM_NAME]: this.config.appId },
        contentType
      )
    );
  }
  async post(endpoint, body = null, contentType) {
    return await this.refreshOnFail(
      (key) => (0, import_request.request)(
        key,
        this.config.batchUrl,
        endpoint,
        "POST",
        body,
        { [import_request.SM_APP_PARAM_NAME]: this.config.appId },
        contentType
      )
    );
  }
  async delete(endpoint, params) {
    return this.refreshOnFail(
      (key) => (0, import_request.request)(key, this.config.batchUrl, endpoint, "DELETE", null, {
        ...params,
        [import_request.SM_APP_PARAM_NAME]: this.config.appId
      })
    );
  }
  /**
   * The main method for transcribing audio files. It takes a url or a buffer and returns a promise with a transcript.
   *
   *
   * @param config TranscribeConfig
   * @returns Promise<RetrieveTranscriptResponse>. A promise that resolves to a transcript.
   */
  async transcribe({
    input,
    fileName,
    transcription_config,
    translation_config,
    output_config,
    summarization_config,
    auto_chapters_config,
    topic_detection_config,
    format = "json-v2"
  }) {
    if (this.config.apiKey === void 0)
      throw new Error("Error: apiKey is undefined");
    const fileOrFetchConfig = "fetch" in input ? input.fetch : input;
    const submitResponse = await this.createJob({
      input: fileOrFetchConfig,
      fileName,
      transcription_config,
      translation_config,
      output_config,
      summarization_config,
      auto_chapters_config,
      topic_detection_config
    });
    if (submitResponse === null || submitResponse === void 0) {
      throw "Error: submitResponse is undefined";
    }
    await (0, import_poll.default)(
      async () => {
        const { job } = await this.getJob(submitResponse.id);
        if (job.status === "rejected") {
          throw job.errors;
        } else if (job.status === "done") {
          return true;
        } else {
          return false;
        }
      },
      3e3,
      // repeat every 3 seconds
      15 * 60 * 1e3
      // 15 minutes timeout
    );
    return await this.getJobResult(submitResponse.id, format);
  }
  async createJob({
    input,
    fileName,
    transcription_config,
    translation_config,
    output_config,
    summarization_config,
    auto_chapters_config,
    topic_detection_config
  }) {
    if (this.config.apiKey === void 0)
      throw new Error("Error: apiKey is undefined");
    const config = {
      type: "transcription",
      transcription_config,
      translation_config,
      output_config,
      summarization_config,
      auto_chapters_config,
      topic_detection_config
    };
    const formData = new FormData();
    if ("url" in input) {
      config.fetch_data = input;
    } else {
      formData.append("data_file", input, fileName);
    }
    formData.append("config", JSON.stringify(config));
    return this.post("/v2/jobs", formData);
  }
  async listJobs(filters = {}) {
    return this.get("/v2/jobs", "application/json", filters);
  }
  async getJob(id) {
    return this.get(`/v2/jobs/${id}`, "application/json");
  }
  async deleteJob(id, force = false) {
    const params = force ? { force } : void 0;
    return this.delete(`/v2/jobs/${id}`, params);
  }
  async getJobResult(jobId, format) {
    const params = { format: format === "text" ? "txt" : format };
    const contentType = format === "json-v2" ? "application/json" : "text/plain";
    return this.get(`/v2/jobs/${jobId}/transcript`, contentType, params);
  }
  async getDataFile(jobId) {
    return this.get(`/v2/jobs/${jobId}/data`, "application/json");
  }
  async getFeatureDiscovery() {
    return this.get("/v1/discovery/features");
  }
}
